loader/src/sources/prepmod/index.js

```diff
@@ -0,0 +1,427 @@
+/**
+ * Load data from PrepMod's SMART Scheduling Links API
+ * https://github.com/smart-on-fhir/smart-scheduling-links/
+ *
+ * PrepMod maintains separate deployments for each customer, so we have to
+ * contact at least one (and sometimes several) hosts in each state.
+ */
+
+const got = require("got");
+const Sentry = require("@sentry/node");
+const { Available, LocationType } = require("../../model");
+const prepmodHostsByState = require("./hosts");
+const { HTTPError } = require("got");
+
+const API_PATH = "/api/smart-scheduling-links/$bulk-publish";
+const MANIFEST_TIMEOUT = 5 * 60 * 1000;
+const HL7_SERVICE_TYPE_SYSTEM =
+  "http://terminology.hl7.org/CodeSystem/service-type";
+const VTRCKS_SYSTEM = "https://cdc.gov/vaccines/programs/vtrcks";
+const CAPACITY_EXTENSION =
+  "http://fhir-registry.smarthealthit.org/StructureDefinition/slot-capacity";
+const BOOKING_DEEP_LINK_EXTENSION =
+  "http://fhir-registry.smarthealthit.org/StructureDefinition/booking-deep-link";
+const PRODUCT_EXTENSION =
+  "http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-product";
+const DOSE_EXTENSION =
+  "http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-dose";
+
+const PRODUCTS_BY_CVX_CODE = {
+  210: "astra_zeneca",
+  207: "moderna",
+  211: "novavax",
+  208: "pfizer",
+  212: "jj",
+};
+
+// Use symbols to link slots -> schedules -> locations to avoid circular
+// references when serializing or logging data.
+const locationReference = Symbol("schedule");
+const scheduleReference = Symbol("schedule");
+
+/**
+ * Parse a Newline-Delimited JSON (NDJSON) document.
+ * @param {string} text
+ * @returns {Array<any>}
+ */
+function parseJsonLines(text) {
+  return text
+    .split("\n")
+    .filter(Boolean)
+    .map((line, index) => {
+      try {
+        return JSON.parse(line);
+      } catch (error) {
+        throw new SyntaxError(`Error parsing line ${index}: ${line}`);
+      }
+    });
+}
+
+/**
+ * Lightweight wrapper for a SMART Scheduling Links API.
+ * This does some basic management around manifest caching and really basic
+ * response parsing, but nothing too fancy.
+ *
+ * SMART SL Docs: https://github.com/smart-on-fhir/smart-scheduling-links/
+ */
+class SmartSchedulingLinksApi {
+  constructor(url) {
+    this.url = url;
+    this.manifest = { time: 0, data: null };
+  }
+
+  async getManifest() {
+    if (
+      !this.manifest.data ||
+      Date.now() - this.manifest.time > MANIFEST_TIMEOUT
+    ) {
+      const data = await got(this.url).json();
+      this.manifest = { time: Date.now(), data };
+    }
+    return this.manifest.data;
+  }
+
+  async *listLocations() {
+    const manifest = await this.getManifest();
+    for (const item of manifest.output) {
+      if (item.type === "Location") {
+        const response = await got(item.url);
+        for (const location of parseJsonLines(response.body)) {
+          yield location;
+        }
+      }
+    }
+  }
```

Comment:
Maybe me trying to abstract too much here, but it seems like it might be useful to have an `async *listItems(itemType)` that applies a simple `item.type` filter that would turn `listLocations`, `listSchedules` and `listSlots` into one-liners.

Although now that I say it, why not pull all three at the same time?

Reply:
Ah, good idea on abstracting the filtering logic to an underlying method! I didn’t pull them all at once because we need to parse all of one type before parsing the next types. (But `getManifest()` has caching, so the separate method calls still only make one HTTP request.)